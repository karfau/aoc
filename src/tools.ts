export const splitLines = (s: string) => s.split(/[\n\r]/).filter(line => line.trim().length > 0);

export const lineToNumbers = (s: string, delim: string | RegExp = '') => s.split(delim).map(c => parseInt(c, 10));

export const runWithStdIn = (fun: (input: string) => any) => {
  require('get-stdin')().then(fun).then(console.log);
};

export type Pair<T> = [T, T]
export type Numbers = ReadonlyArray<number>;

