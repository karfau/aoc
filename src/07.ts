import {runWithStdIn, splitLines} from './tools';
import {difference, flatten, sum, uniq} from 'lodash';

export type Program = {
  w: number;
  c: string[];
  s?: number;
}

export type Tower = {
  [name: string]: Program
}

export function findRoot(lines: ReadonlyArray<string>): string {
  const children: string[] = [];
  return difference(lines.map(line => {
    const [head, ...below] = line.split(/[ (\d)\->,]+/);
    children.push(...below);
    return head;
  }), children)[0];
}

export const tower = (lines: ReadonlyArray<string>): Tower => {
  return lines.reduce((tower, line) => {
    const [head, w, ...c] = line.split(/[\s()\->,]+/);
    tower[head as string] = {w: parseInt(w), c: c.length === 1 && c[0] === '' ? [] : c};
    return tower;
  }, {} as Tower);
};

export const tasks = (t: Tower, root: string): string[][] => {
  const result: string[][] = [];
  let current: string[] = [root];
  do {
    const next = flatten(current.map(c => c in t ? t[c].c : []));
    result.push(current);
    current = next;
  } while (current.length > 0);
  return result;
};

export type Sums = {[key: string]: [number, number[]]}

export const sums = (t:Tower, tasks: string[][]): Sums => {
  return [...tasks].reverse().reduce((result, tasklevel, lvl) => {
    tasklevel.reduce((r, task) => {
      const cw = t[task].c.map(c => r[c][0]);
      r[task] = [sum(cw) + t[task].w, cw];
      if (uniq(cw).length>1){
        console.log(tasks.length - lvl, cw, t[task].c.map(c => [c, t[c].w, ...r[c]]));
      }
      return r
    }, result);
    return result;
  }, {} as Sums)
};

export const main = (input: string) => {
  const lines = splitLines(input);
  const root = findRoot(lines);
  const t = tower(lines);
  sums(t, tasks(t, root))
};

if (require.main === module) runWithStdIn(main);
