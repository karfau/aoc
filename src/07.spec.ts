import {splitLines} from './tools';
import {findRoot, Sums, sums, tasks, Tower, tower} from './07';

const INPUT = `
pbga (66)
xhth (57)
ebii (61)
havc (66)
ktlj (57)
fwft (72) -> ktlj, cntj, xhth
qoyq (66)
padx (45) -> pbga, havc, qoyq
tknk (41) -> ugml, padx, fwft
jptl (61)
ugml (68) -> gyxo, ebii, jptl
gyxo (61)
cntj (57)
`;

describe('day 7', function () {
  describe('findRoot()', function () {
    it(`"should find tknk`, function () {
      expect(findRoot(splitLines(INPUT))).toEqual('tknk');
    });
  });
  describe('tree()', function () {
    it(`"should create tree for`, function () {
      expect(tower(splitLines(INPUT))).toMatchSnapshot();
    });
  });
  describe('tasks()', function () {
    it(`"should create tasks for one level`, function () {
      expect(tasks({root:{w: NaN,c:['a', 'b']}}, 'root'))
        .toEqual([['root'], ['a', 'b']]);
    });
    it(`"should create tasks for two levels`, function () {
      expect(tasks(
        {
          root:{w: NaN,c:['a1', 'b1']},
          a1:{w: NaN,c:['c2']}
        }, 'root')).toEqual([['root'], ['a1', 'b1'], ['c2']]);
    });
    it(`"should create tasks for est input`, function () {
      expect(tasks(tower(splitLines(INPUT)), 'tknk')).toMatchSnapshot();
    });
  });
  describe('sums()', function () {
    let t: Tower;
    beforeAll(() => {
      t = tower(splitLines(INPUT));
    });
    it(`"should create sum for single`, function () {
      const expected: Sums = {jptl:[61, []]};
      expect(sums(t, tasks(t, 'jptl'))).toEqual(expected);
    });
    it(`"should create sums for ugml`, function () {
      const expected: Sums = {ugml: [251,[61, 61, 61]], gyxo: [61, []], ebii:[61, []], jptl:[61, []]};
      expect(sums(t, tasks(t, 'ugml'))).toEqual(expected);
    });
    it(`"should create sums for test input`, function () {
      const expected: Sums = {ugml: [251,[61, 61, 61]], gyxo: [61, []], ebii:[61, []], jptl:[61, []]};
      expect(sums(t, tasks(t, 'tknk'))['tknk']).toEqual([778, [251, 243, 243]]);
    });
  });
});
