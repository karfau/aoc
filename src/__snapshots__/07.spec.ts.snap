exports[`day 7 tasks() "should create tasks for est input 1`] = `
Array [
  Array [
    "tknk",
  ],
  Array [
    "ugml",
    "padx",
    "fwft",
  ],
  Array [
    "gyxo",
    "ebii",
    "jptl",
    "pbga",
    "havc",
    "qoyq",
    "ktlj",
    "cntj",
    "xhth",
  ],
]
`;

exports[`day 7 tree() "should create tree for 1`] = `
Object {
  "cntj": Object {
    "c": Array [],
    "w": 57,
  },
  "ebii": Object {
    "c": Array [],
    "w": 61,
  },
  "fwft": Object {
    "c": Array [
      "ktlj",
      "cntj",
      "xhth",
    ],
    "w": 72,
  },
  "gyxo": Object {
    "c": Array [],
    "w": 61,
  },
  "havc": Object {
    "c": Array [],
    "w": 66,
  },
  "jptl": Object {
    "c": Array [],
    "w": 61,
  },
  "ktlj": Object {
    "c": Array [],
    "w": 57,
  },
  "padx": Object {
    "c": Array [
      "pbga",
      "havc",
      "qoyq",
    ],
    "w": 45,
  },
  "pbga": Object {
    "c": Array [],
    "w": 66,
  },
  "qoyq": Object {
    "c": Array [],
    "w": 66,
  },
  "tknk": Object {
    "c": Array [
      "ugml",
      "padx",
      "fwft",
    ],
    "w": 41,
  },
  "ugml": Object {
    "c": Array [
      "gyxo",
      "ebii",
      "jptl",
    ],
    "w": 68,
  },
  "xhth": Object {
    "c": Array [],
    "w": 57,
  },
}
`;
